### PR TITLE
Add support for API timeouts to auth-next

### PR DESCRIPTION
**Pull request by avolkovi**
Hey there! So you want to contribute to a Firebase SDK? 
Before you file this pull request, please read these guidelines:

### Discussion

  * Read the contribution guidelines (CONTRIBUTING.md).
  * If this has been discussed in an issue, make sure to link to the issue here. 
    If not, go file an issue about this **before creating a pull request** to discuss.

### Testing

  * Make sure all existing tests in the repository pass after your change.
  * If you fixed a bug or added a feature, add a new test to cover your code.

### API Changes

  * At this time we cannot accept changes that affect the public API.  If you'd like to help 
    us make Firebase APIs better, please propose your change in an issue so that we 
    can discuss it together.

### CODE REVIEW

File: packages-exp/auth-exp/src/api/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    responseKey: 'response-value'
+  };
+
+  context('with regular requests', () => {
+    beforeEach(mockFetch.setUp);
+    afterEach(mockFetch.tearDown);
+
+    it('should set the correct request, method and HTTP Headers', async () => {
+      const mock = mockEndpoint(Endpoint.SIGN_UP, serverResponse);
+      const response = await performApiRequest<typeof request, typeof serverResponse>(mockAuth, HttpMethod.POST, Endpoint.SIGN_UP, request);
```
Comment (samhorlbeck) [inline packages-exp/auth-exp/src/api/index.test.ts:46]: nit: line length

File: packages-exp/auth-exp/src/api/index.ts [hunk 2]
```diff
@@ -53,6 +53,8 @@ export enum Endpoint {
   WITHDRAW_MFA = '/v2/accounts/mfaEnrollment:withdraw'
 }
 
+export const DEFAULT_API_TIMEOUT = 30_000;
```
Comment (samhorlbeck) [inline packages-exp/auth-exp/src/api/index.ts:56]: Should this have _MS for clarity?

File: packages-exp/auth-exp/src/api/index.ts [hunk 3]
```diff
@@ -77,7 +79,9 @@ export async function performApiRequest<T, V>(
         referrerPolicy: 'no-referrer',
         ...body
       }
-    );
+    ), new Promise((_, reject) =>
+      setTimeout(() => reject(AUTH_ERROR_FACTORY.create(AuthErrorCode.TIMEOUT, { appName: auth.name })), DEFAULT_API_TIMEOUT)
```
Comment (samhorlbeck) [inline packages-exp/auth-exp/src/api/index.ts:83]: line length

File: packages-exp/auth-exp/src/api/index.ts [hunk 4]
```diff
@@ -88,7 +92,8 @@ export async function performApiRequest<T, V>(
       } else {
         // TODO probably should handle improperly formatted errors as well
         // If you see this, add an entry to SERVER_ERROR_MAP for the corresponding error
-        throw new Error(`Unexpected API error: ${json.error.message}`);
+        console.error(`Unexpected API error: ${json.error.message}`);
```
Comment (samhorlbeck) [inline packages-exp/auth-exp/src/api/index.ts:95]: Is this meant to be here?

File: packages-exp/auth-exp/src/api/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+describe('performApiRequest', () => {
+  const request = {
+    requestKey: 'request-value'
+  };
+
+  const serverResponse = {
+    responseKey: 'response-value'
+  };
+
+  context('with regular requests', () => {
```
Comment (Feiyang1) [inline packages-exp/auth-exp/src/api/index.test.ts:40]: Is context an alias to describe?

Comment (avolkovi) [reply packages-exp/auth-exp/src/api/index.test.ts:40]: yes, it's just an alias that mocha inherit from rspec https://lmws.net/describe-vs-context-in-rspec

TLDR describe encompasses the functionality you want to test, context groups similar setup blocks for the same functionality

File: packages-exp/auth-exp/src/api/index.ts (hunk 4 above)
Comment (avolkovi) [reply packages-exp/auth-exp/src/api/index.ts:95]: I added the console.error since we never expect to see these, it would help in debugging since we mask json.error.message with an error from the map (in this case, internal-error)

File: packages-exp/auth-exp/src/api/index.test.ts (hunk 5 above)
Comment (Feiyang1) [reply packages-exp/auth-exp/src/api/index.test.ts:40]: Nice!
